### PR TITLE
fix wait for deposit failure and junk proposals and attestations before sync

### PR DESCRIPTION
**Pull request by shayzluf**
# Fix beacon node and validator failures before initial sync
starting beacon node and validator while following the steps in https://prylabs.net/participate results in these unwanted behaviors:
1. beacon node RPC service didn't start - validator die
2. beacon node RPC service started but sync didn't start (state==nil) - beacon node WaitForActivation cant handle it and die
3. beacon node RPC service started but sync started and got to a point were validator is activated but initial sync didn't finish - validator starts performing its duties and submits attestations and proposals on old state

this pr handles 2 and 3 we should consider blocking propose and attest on the beacon node while sync service is erroneous

### CODE REVIEW

File: beacon-chain/rpc/validator_server.go [hunk 1]
```diff
@@ -73,6 +75,23 @@ func (vs *ValidatorServer) WaitForActivation(req *pb.ValidatorActivationRequest,
 		}
 	}
 }
+// WaitTillSync checks if syncservice status is synched and sends message on stream when the service is synced.
+func (vs *ValidatorServer) WaitTillSync(_ *ptypes.Empty,stream pb.ValidatorService_WaitTillSyncServer) error{
```
Comment (nisdas) [inline beacon-chain/rpc/validator_server.go]: why does this need to be a stream ?

Comment (shayzluf) [reply beacon-chain/rpc/validator_server.go]: it doesn't need to be a stream but we might later need this stream to know if we went out of sync and not just before initial sync

Comment (nisdas) [reply beacon-chain/rpc/validator_server.go]: this breaks on the client side , the moment we are synced. So there is no way to know if we went out of sync or not. I dont see why does this need to be a stream

Comment (terencechain) [reply beacon-chain/rpc/validator_server.go]: +1

Comment (shayzluf) [reply beacon-chain/rpc/validator_server.go]: changed

File: beacon-chain/rpc/validator_server.go [hunk 2]
```diff
@@ -74,6 +76,15 @@ func (vs *ValidatorServer) WaitForActivation(req *pb.ValidatorActivationRequest,
 	}
 }
 
+// WaitTillSync checks if syncservice status is synched and sends message on stream when the service is synced.
+func (vs *ValidatorServer) WaitTillSync(ctx context.Context,_ *ptypes.Empty) (*pb.SyncedResponse, error){
+	if err := vs.syncService.Status();err==nil{
+		return &pb.SyncedResponse{Synced:true},nil
```
Comment (nisdas) [inline beacon-chain/rpc/validator_server.go]: we dont need a special proto message for this. just return the error from `Status`. We can then handle the error client side. `if err == nil {break} ` or something

File: validator/client/validator.go [hunk 3]
```diff
@@ -118,6 +118,28 @@ func (v *validator) WaitForActivation(ctx context.Context) error {
...
+func (v *validator) WaitTillSync(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "validator.WaitTillSync")
+	defer span.End()
+	for {
+		log.Info("Waiting for beacon node to sync")
+		// If context is canceled we stop the loop.
+		if ctx.Err() == context.Canceled {
+			return fmt.Errorf("context has been canceled so shutting down the loop: %v", ctx.Err())
+		}
+		res, _ := v.validatorClient.WaitTillSync(ctx, &ptypes.Empty{})
```
Comment (nisdas) [inline validator/client/validator.go]: we handle the error here , if its nil, then only we break

File: validator/client/validator.go [hunk 4]
```diff
@@ -118,6 +118,28 @@ func (v *validator) WaitForActivation(ctx context.Context) error {
...
+		if res != nil && res.Synced {
+			log.Info("Beacon node is Synched!")
```
Comment (nisdas) [inline validator/client/validator.go]: `is Synced`

File: beacon-chain/rpc/validator_server.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/rpc/validator_server.go]: maybe i will add the sync error separately

Comment (shayzluf) [reply beacon-chain/rpc/validator_server.go]: i am not able to return primitive types in protobuf

File: validator/client/validator.go (hunk 4 above)
Comment (shayzluf) [reply validator/client/validator.go]: fixed

**Pull request closed without merging**